### PR TITLE
8316421: libjava should load shell32.dll eagerly

### PR DESCRIPTION
**Pull request by djelinski**
Please review this patch that makes java.dll load shell32.dll earlier. Delay-loading requires some additional code (delayimp.lib), and offers no benefits since we always load shell32 during JVM startup.

Other than removing the delayload clause, the patch also cleans up the `getHomeFromShell32` method:
- the WinXP code path is removed. The documentation states that [the older function just calls the new one with translated parameters](https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetfolderpatha).
- `CoTaskMemFree` is now called if `SHGetKnownFolderPath` fails. This is suggested by the documentation, but probably never needed in practice.

This change shouldn't have any observable effect on behavior on any of the supported operating systems. It reduced the size of java.dll by 2KB on my machine.
No new tests. Existing tier1-2 tests continue to pass.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316421](https://bugs.openjdk.org/browse/JDK-8316421): libjava should load shell32.dll eagerly (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [1fa01288](https://git.openjdk.org/jdk/pull/15789/files/1fa012885ae50f649d6038db3562b930adb9df2f)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer) ⚠️ Review applies to [1fa01288](https://git.openjdk.org/jdk/pull/15789/files/1fa012885ae50f649d6038db3562b930adb9df2f)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15789/head:pull/15789` \
`$ git checkout pull/15789`

Update a local copy of the PR: \
`$ git checkout pull/15789` \
`$ git pull https://git.openjdk.org/jdk.git pull/15789/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15789`

View PR using the GUI difftool: \
`$ git pr show -t 15789`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15789.diff">https://git.openjdk.org/jdk/pull/15789.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15789#issuecomment-1723649038)

### CODE REVIEW

**Comment by djelinski** [comment]
Well yes, if `SHGetKnownFolderPath` were available in Win2k, delayload would probably have been removed in 3b0f760747ae66cdd343a5e1d623af65e089b442.

Oracle docs state that JDK17 only supports Win8 and newer:
https://www.oracle.com/java/technologies/javase/products-doc-jdk17certconfig.html

We recently integrated a change that requires Vista or newer:
[JDK-8302659](https://bugs.openjdk.org/browse/JDK-8302659)

During compilation we enable APIs from Win8:
https://github.com/openjdk/jdk/blob/2e2d49c76d7bb43a431b5c4f2552beef8798258b/make/autoconf/flags-cflags.m4#L494

based on the above, I believe we can ignore WinXP compatibility here.

**Comment by JornVernee** [comment]
> based on the above, I believe we can ignore WinXP compatibility here.

I tend to agree.

> During compilation we enable APIs from Win8:

This is interesting. I think that means the `define` at the start of src/java.base/windows/native/libjava/java_props_md.c can also be removed

**Comment by djelinski** [comment]
Hmm, indeed `VER_PLATFORM_WIN32_WINDOWS` was not used... Removed now. Was that the `define` you were referring to?

**Comment by JornVernee** [comment]
> Hmm, indeed `VER_PLATFORM_WIN32_WINDOWS` was not used... Removed now. Was that the `define` you were referring to?

Err, sorry, I was talking about the `_WIN32_NT` define, but that was actually already removed in https://github.com/openjdk/jdk/commit/4fd79a6ad2683e4863bd4e311cb01cbc30ebf57f#diff-e17e04444071ba72ceac86556f300573321b4317b88bbd694e55f51031b7e3db

**Review by erikj79 — APPROVED** [review]
Build change looks good.

**Comment by TheShermanTanker** [comment]
Seems like JDK-8288293 no longer has to rely on delay load failure hooks to bypass gcc's lack of Structured Exception Handling :)

Is the ole32.lib dependency related to CoTaskMemFree?

**Comment by djelinski** [comment]
Yes, ole32.lib was needed for CoTaskMemFree. Without it the library wouldn't link.

**Review by TheShermanTanker — APPROVED** [review]
Looks Great!

File: src/java.base/windows/native/libjava/java_props_md.c [hunk 1]
```diff
@@ -214,38 +210,13 @@ getHomeFromShell32()
...
-             * [SHGetKnownFolderPath] call to avoid MAX_PATH length limitation.
-             * Shell32.dll (version 6.0.6000 or later)
-             */
-            hr = SHGetKnownFolderPath(&FOLDERID_Profile, KF_FLAG_DONT_VERIFY, NULL, &u_path);
-        } __except(EXCEPTION_EXECUTE_HANDLER) {
-            /* Exception: no [SHGetKnownFolderPath] entry */
-            hr = E_FAIL;
-        }
+        WCHAR *tmpPath = NULL;
+        hr = SHGetKnownFolderPath(&FOLDERID_Profile, KF_FLAG_DONT_VERIFY, NULL, &tmpPath);
```
Comment (JornVernee) [inline src/java.base/windows/native/libjava/java_props_md.c:214]: I'd say inline the variable declaration here as well
```suggestion
        HRESULT hr = SHGetKnownFolderPath(&FOLDERID_Profile, KF_FLAG_DONT_VERIFY, NULL, &tmpPath);
```

**Review by JornVernee — APPROVED** [review]
LGTM

File: src/java.base/windows/native/libjava/java_props_md.c (hunk 1 above)
Comment (JornVernee) [reply src/java.base/windows/native/libjava/java_props_md.c:214]: (And you will have to remove the declaration from the line above)

Comment (djelinski) [reply src/java.base/windows/native/libjava/java_props_md.c]: Done

**Review by JornVernee — APPROVED** [review]

**Comment by djelinski** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**